sql/ttl: improve TTL replan decision logic

spilchen opened the pull request:
Replace calculatePlanGrowth with detectNodeAvailabilityChanges to make TTL job replanning less sensitive to span changes. The new logic focuses specifically on detecting when nodes become unavailable rather than reacting to all plan differences.

The previous implementation would trigger replans for span splits/merges that don't actually indicate beneficial restart scenarios. The new approach only considers missing nodes from the original plan, which typically indicates node failures where work redistribution would benefit from restarting the job. It also supports a stability window so that replan decisions need to fire consecutively. This should help aleviate changes in plans due to range cache issues.

Fixes #150343

Epic: none
Release note (ops change): The 'sql.ttl.replan_flow_threshold' may have been set to 0 to work around the TTL replanner being too sensitive. This fix will alleviate that and any instance that had set replan_flow_threshold to 0 can be reset back to the default.

fqazi reviewed (changes requested):
> Reviewed 1 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OW1EaoM2MKRKrAe8wT6:b7n0ed0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
> ```Go
> 
> 	var frac float64
> 	if changed > 0 {
> ```

There are other reasons where the nodes may shift in a plan, since it could be also a function of the oracle that selects replicas for each span (i.e. the latency varying, lease holders changing, etc..):  https://github.com/cockroachdb/cockroach/blob/19fcdb66fdaf199077c5dd00c45c7b561c7cb398/pkg/sql/physicalplan/replicaoracle/oracle.go#L260

Do we want to be more concrete about node deaths by doing a liveness check? Using the cached reader to detect node death: https://github.com/cockroachdb/cockroach/blob/19fcdb66fdaf199077c5dd00c45c7b561c7cb398/pkg/sql/sqlliveness/sqlliveness.go#L116


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OW1Ssi4BdimopQzQVSW:b-hghvhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
> There are other reasons where the nodes may shift in a plan, since it could be also a function of the oracle that selects replicas for each span (i.e. the latency varying, lease holders changing, etc..):

Actually, maybe the logic in the current revision is a good middle ground. It might be a good idea to re-plan if that many leases are transferred to new nodes. Ultimately what we want here is to use as much CPU power for the TTL processors as makes sense, without being overly sensitive to the normal movement of ranges.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OW1T_PVAJkuXJJvbnMZ:bsd7d2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

> There are other reasons where the nodes may shift in a plan, since it could be also a function of the oracle that selects replicas for each span (i.e. the latency varying, lease holders changing, etc..):

Actually, maybe the logic in the current revision is a good middle ground. It might be a good idea to re-plan if that many leases are transferred to new nodes. Ultimately what we want here is to use as much CPU power for the TTL processors as makes sense, without being overly sensitive to the normal movement of ranges.
</blockquote></details>

I'm more worried about other factors, since it will only use 10 ranges from the lease holder. After that it will be replicas in some sorted order based on things like latency,. Not sure how stable it will be or how much of a change will be observed on a typical cluster underload. Just raising it as a concern.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OW1VtkEEgxxrAIIuCbx:b-zi645q) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I'm more worried about other factors, since it will only use 10 ranges from the lease holder. After that it will be replicas in some sorted order based on things like latency,. Not sure how stable it will be or how much of a change will be observed on a typical cluster underload. Just raising it as a concern.
</blockquote></details>

Yeah, it would be nice if we knew a ballpark estimation of how frequently the other replicas move and would cause a plan change here. We might be able to infer something from looking at [KV_DISTRIBUTION logs](https://us5.datadoghq.com/logs?query=source%3Acockroachdb%20env%3Aproduction%20%40channel%3AKV_DISTRIBUTION%20%40cluster_type%3ADEDICATED%20replica%20-%22error%20processing%22&agg_m=count&agg_m_source=base&agg_t=count&cols=host%2Cservice&messageDisplay=inline&refresh_mode=sliding&storage=flex_tier&stream_sort=desc&viz=stream&from_ts=1753376301921&to_ts=1753462701921&live=true)


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OW1WH5r4soXixMlFlPi:bmqtcrn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

Yeah, it would be nice if we knew a ballpark estimation of how frequently the other replicas move and would cause a plan change here. We might be able to infer something from looking at [KV_DISTRIBUTION logs](https://us5.datadoghq.com/logs?query=source%3Acockroachdb%20env%3Aproduction%20%40channel%3AKV_DISTRIBUTION%20%40cluster_type%3ADEDICATED%20replica%20-%22error%20processing%22&agg_m=count&agg_m_source=base&agg_t=count&cols=host%2Cservice&messageDisplay=inline&refresh_mode=sliding&storage=flex_tier&stream_sort=desc&viz=stream&from_ts=1753376301921&to_ts=1753462701921&live=true)
</blockquote></details>

Another approach is to add a stability window. Basically, trigger replanning if the plan changes persist across multiple checks. For example, 2-3 consecutive replan evaluations will trigger a replan. That may filter out transient lease movements.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OW1dxN37aMbOQoj9T6O:b-bpo18z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Another approach is to add a stability window. Basically, trigger replanning if the plan changes persist across multiple checks. For example, 2-3 consecutive replan evaluations will trigger a replan. That may filter out transient lease movements.
</blockquote></details>

I looked through the debug zip from the support case, I did see a lot of lease transfers in the cluster.  On the order of 1000s per day. This makes me think that checking the sql liveness check may be the way to go then. What do you think?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OW1e_n4A9OgvAZon7Hg:b-p09999) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I looked through the debug zip from the support case, I did see a lot of lease transfers in the cluster.  On the order of 1000s per day. This makes me think that checking the sql liveness check may be the way to go then. What do you think?
</blockquote></details>

Or maybe a hybrid approach? Re-plan instantly if we detect dead nodes otherwise, we some interval / stability window based thing?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OW1ezSf5Zd2Nn3bizUN:b-96bkkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Or maybe a hybrid approach? Re-plan instantly if we detect dead nodes otherwise, we some interval / stability window based thing?
</blockquote></details>

*with some interval


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OW1hDc-5jkJ7ETRYpMz:b-ni2kv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
> I looked through the debug zip from the support case, I did see a lot of lease transfers in the cluster. On the order of 1000s per day. This makes me think that checking the sql liveness check may be the way to go then. What do you think?

Well, the number of lease transfers itself is not the information we need. The plan would only change if there were enough lease transfers to nodes that previously were not leaseholders at all. I imagine that would be significantly less frequent.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OW1zmVe9zLdmLk0r1EU:b-xsq6qf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

> I looked through the debug zip from the support case, I did see a lot of lease transfers in the cluster. On the order of 1000s per day. This makes me think that checking the sql liveness check may be the way to go then. What do you think?

Well, the number of lease transfers itself is not the information we need. The plan would only change if there were enough lease transfers to nodes that previously were not leaseholders at all. I imagine that would be significantly less frequent.
</blockquote></details>

Ah, good point. It looked like most of the rebalancing are between the same set of (healthy?) nodes. So, it’s likely the differences in the spans at each processor that is probably attributing to the replan, which should be addressed now. I can look at adding a stability window to try and smooth this out.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OW1EaoM2MKRKrAe8wT5:-OWLHJN1B85Xq9Ho4gnH:b3nsa64) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a00c1b64db81af5abbccb3e60d1e7a677af526e/pkg/sql/ttl/ttljob/ttljob.go#L557)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Ah, good point. It looked like most of the rebalancing are between the same set of (healthy?) nodes. So, it’s likely the differences in the spans at each processor that is probably attributing to the replan, which should be addressed now. I can look at adding a stability window to try and smooth this out.
</blockquote></details>

I added the stability window.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
looks good so far!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-:-OWLsZfg9wZHw63A5O5v:b-xo2sg4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`-- commits` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OWLpgAS66Y4ESTpIE16:-OWLpgAS66Y4ESTpIE17:b-52hnlj):*
super nit: let's use "ops change" for the category; that's the general bucket that TTL falls under.
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OWLsCuM55kqkTlsXhLW:-OWLsCuM55kqkTlsXhLX:b-ec79g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7044ae895d78d4a0c88f2926568fd42676b2c3a/pkg/sql/ttl/ttljob/ttljob.go#L59)):*
> ```Go
> )
> 
> var replanStabilityWindow = settings.RegisterIntSetting(
> ```

out of curiosity, do you know how frequently the plan gets rechecked? that could help decide what a good default stability window is.
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OWLq2raDqlg68OKumtV:-OWLq2rb8j3TJ4Gw9Zym:b-kxemm3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7044ae895d78d4a0c88f2926568fd42676b2c3a/pkg/sql/ttl/ttljob/ttljob.go#L78)):*
> ```Go
> 
> 	// consecutiveReplanDecisions tracks how many consecutive times replan was deemed necessary.
> 	consecutiveReplanDecisions atomic.Int64
> ```

nit: let's make this be of the type `*atomic.Int64` so it's easier to remember how to pass it as a parameter.
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 531 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OWLqnwP4bFvc3ByMFpf:-OWLqnwP4bFvc3ByMFpg:b-yrg08z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7044ae895d78d4a0c88f2926568fd42676b2c3a/pkg/sql/ttl/ttljob/ttljob.go#L531)):*
> ```Go
> 	consecutiveReplanDecisions *atomic.Int64,
> 	stabilityWindowFn func() int64,
> 	thresholdFn func() float64,
> ```

what's the advantage of passing in these two parameters as functions, rather than as just the numerical values?
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 548 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OWLr93G65Qa78nu_qnK:-OWLr93G65Qa78nu_qnL:b-19nr8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7044ae895d78d4a0c88f2926568fd42676b2c3a/pkg/sql/ttl/ttljob/ttljob.go#L548)):*
> ```Go
> 		}
> 
> 		// If stability window is 1, replan immediately (current behavior)
> ```

nit: let's remove "current behavior" here, and make sure to add periods at the ends of sentences.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/ttl: improve TTL replan decision logic

Replace calculatePlanGrowth with detectNodeAvailabilityChanges to make
TTL job replanning less sensitive to span changes. The new logic focuses
specifically on detecting when nodes become unavailable rather than
reacting to all plan differences.

The previous implementation would trigger replans for span splits/merges
that don't actually indicate beneficial restart scenarios. The new
approach only considers missing nodes from the original plan, which
typically indicates node failures where work redistribution would benefit
from restarting the job. It also supports a stability window so that
replan decisions need to fire consecutively. This should help eleviate
changes in plans due to range cache issues.

Fixes #150343

Epic: none
Release note (ops change): The 'sql.ttl.replan_flow_threshold' may
have been set to 0 to work around the TTL replanner being too sensitive.
This fix will alleviate that and any instance that had set
replan_flow_threshold to 0 can be reset back to the default.

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @rafiss)

___
*[`-- commits` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OWLpgAS66Y4ESTpIE16:-OWMPAGB8u6eGHX8w8sb:b-896fix):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: let's use "ops change" for the category; that's the general bucket that TTL falls under.
</blockquote></details>

Done.
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OWLsCuM55kqkTlsXhLW:-OWMNZY96jIbj0MSTr5W:b3u443o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7044ae895d78d4a0c88f2926568fd42676b2c3a/pkg/sql/ttl/ttljob/ttljob.go#L59)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

out of curiosity, do you know how frequently the plan gets rechecked? that could help decide what a good default stability window is.
</blockquote></details>

By default it's every 2 minutes. So, it'll take about 4 minutes after a node goes down, or comes up, for the TTL job to get replanned.
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 531 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-OWLqnwP4bFvc3ByMFpf:-OWMO-0--Vh5Op6Gge3D:b-4bpuxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7044ae895d78d4a0c88f2926568fd42676b2c3a/pkg/sql/ttl/ttljob/ttljob.go#L531)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

what's the advantage of passing in these two parameters as functions, rather than as just the numerical values?
</blockquote></details>

It allows you to change the value of the settings in the middle of a TTL job and pick up the new setting.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! great job nailing down a better check here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150771#-:-OWMarPn2P5c_khCyYXC:b-q6ydhr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)



<!-- Sent from Reviewable.io -->
